<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [188]:
# импортируем библиотеки
import pandas as pd
import psycopg2

In [189]:
# параметры для подключения к SQL
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [190]:
# создаем соединение с заданными параметрами
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [191]:
# запрос для расчета количества вакансий в базе
query_3_1 = f'''select count(*)
                from public.vacancies
            '''
    
# выводим данные по запросу            
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\3953600379.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [192]:
# запрос для расчета количества работодателей в базе
query_3_2 = f'''select count(*)
                from public.employers
            '''
            

# выводим данные по запросу             
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\477333069.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [193]:
# запрос для расчета количества регионов в базе
query_3_3 = f'''select count(*)
                from public.areas
            '''
            

# выводим данные по запросу             
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\1260545241.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [194]:
# запрос для расчета количества сфер деятельности в базе
query_3_4 = f'''select count(*)
                from public.industries
            '''
            

# выводим данные по запросу             
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\3893265048.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

### Проведя предварительный анализ данных, можно сделать следующие выводы:
#### - в базе содержатся данные по 49197 вакансиям 23501 работодателей, расположенных в 1362 регионах
#### - в базе представлены вакансии для 294 сфер деятельности
#### - HH имеет большую базу данных, для исследования который требуется применять специальные программы, такие, например, как SQL, т.к. такой объем данных невозможно проанализировать вручную

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [195]:
# запрос для расчета количества вакансий в каждом регионе
query_4_1 = f'''SELECT 
                    count(distinct v.name) cnt,
                    a.name area
                from public.vacancies v
                left join public.areas a on v.area_id=a.id
                group by a.name
                order by count(distinct v.name) desc
                limit 5
            '''
            

# выводим данные по запросу             
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2321350828.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             area
0  3894           Москва
1  2143  Санкт-Петербург
2  1501      Новосибирск
3  1497            Минск
4  1421           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [196]:
# запрос для определения количества вакансий, у которых заполнено хотя бы одно значение зарплаты (от/до)
query_4_2 = f'''SELECT 
                COUNT(*)
                from public.vacancies
                WHERE salary_from is not null or salary_to is not null
            '''
            

# выводим данные по запросу             
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2016554713.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [197]:
# запрос для расчета среднего значения для нижней (от) и верхней (до) границы зарплатной вилки
query_4_3 = f'''SELECT 
                AVG(salary_from),
                AVG(salary_to)
                from public.vacancies
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\613405266.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg            avg
0  71064.657901  110536.741923

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [198]:
# запрос для расчета количества вакансий для каждого сочетания рабочего графика и типа трудоустройства
query_4_4 = f'''SELECT
                schedule,
                employment,
                count(*)
                from public.vacancies
                group by schedule, employment
                order by 3 desc
            '''
     

# выводим данные по запросу             
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2950301942.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [199]:
# запрос для расчета количества вакансий в зависимости от требуемого опыта работы
query_4_5 = f'''SELECT
                DISTINCT(experience),
                COUNT(id)
                FROM public.vacancies v
                GROUP BY experience
                ORDER BY 2 ASC
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\741832106.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

### Проведя детальный анализ базы данных, можно сделать следующие выводы:
#### - самое большое количество  вакансий представлено в двух самых крупных городах Росии - Москве и Санкт-Петербурге, на втором месте после России идет Белоруссия и на третьем Казахстан;
#### - только у половины из всех представленных вакансий заполнено одно из двух полей с зарплатой (24073 из 49197 вакансий). Думаю эти вакансии будут более интересны для соискателей, т.к. значение зарплаты очень важно при поиске работы
#### - средние значения нижней и верхней границы зарплатной вилки отличаются почти на 50%. вообще данные показатели мало о чем могут сказать, считаю более целесообразным средние значения нижней и верхней границы считать в зависимости от сферы деятельности или типа вакансии
#### - наибольшее количество всех пресдтавленных вакансий предусматривают полную занятость и полный рабочий день, также мы видим, что набирает популярность удаленная работа при полной занятости. Это может быть связано с тем, что набирают популярность профессии, для которых не требуется постоянное посещение офиса работодателя.
#### - самое большое количество вакансий представлено для требуемого опыта работы от 1 года до 3 лет, самое низкое количество вакансий для опыта работы более 6 лет. Думаю это связано с тем, что работодателю важнее, чтобы соискатель имел определенные навыки работы, а не высокий трудовой стаж, т.к. зачастую люди с небольшим опытом работы более "гибкие" в работе.

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [200]:
# запрос для расчета количества вакансий по работодателям
query_5_1 = f'''(select 
                e.name, 
                Count(v.id)
                from public.employers e
                join public.vacancies v on v.employer_id = e.id
                group by 1
                order by 2 desc
                limit 1)

                union all

                (select
                ee.name, 
                Count(vv.id)
                from public.employers ee
                join public.vacancies vv on vv.employer_id = ee.id
                group by 1
                order by 2 desc
                offset 4 limit 1)
            '''


# выводим данные по запросу 
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\779523981.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [201]:
# запрос для расчета количества работодателей и вакансий в каждом регионе
query_5_2 = f'''select
                a.name,
                COUNT(v.id) total_vacancies,
                COUNT(e.id) total_employers
                FROM areas a
                join employers e ON a.id = e.area
                left join vacancies v ON a.id = v.area_id
                GROUP BY a.name
                order by 2, 3 desc
            '''
 

# выводим данные по запросу             
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\1659687854.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  total_vacancies  total_employers
0                Россия                0              410
1             Казахстан                0              207
2    Московская область                0               75
3    Краснодарский край                0               19
4    Ростовская область                0               18
..                  ...              ...              ...
649         Новосибирск          1149438          1149438
650              Алматы          1364132          1364132
651               Минск          2354880          2354880
652     Санкт-Петербург          6320667          6320667
653              Москва         31272712         31272712

[654 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [202]:
# запрос для расчета количества регионов, в которых каждый работодатель публикует свои вакансии
query_5_3 = f'''SELECT
                e.name, 
                count(distinct v.area_id)
                FROM public.employers e 
                LEFT JOIN public.vacancies v ON e.id = v.employer_id
                GROUP BY e.name
                ORDER BY 2 DESC
            '''
 

# выводим данные по запросу             
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\1286625278.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
23170                  СДЕЛКА      0
23171            Alandr Group      0
23172                     СДК      0
23173      Lemon Land Lombard      0
23174                  Махина      0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [203]:
# запрос для подсчета количества работаодателей, у которых не указана сфера деятельности
query_5_4 = f'''SELECT
                count(e.id)
                FROM employers e 
                LEFT JOIN employers_industries ei ON e.id = ei.employer_id
                WHERE ei.industry_id is null
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2967760613.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [204]:
# запрос для поиска работодателя, у которого указано 4 сферы деятельности и который находится на 3 месте в алфавитном порядке
query_5_5 = f'''SELECT
                e.name,
                count(ei.industry_id)
                FROM employers e 
                LEFT JOIN employers_industries ei ON e.id = ei.employer_id
                GROUP by e.name
                HAVING count(ei.industry_id) = '4' -- ищем работодателей, у которых указано 4 сферы деятельности
                order by 1 -- сортируем в алфавитном порядке
                offset 2 LIMIT 1
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\741805940.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [205]:
# запрос для расчета количества работодателей, у которых сфера деятельности "Разработка программного обеспечения"
query_5_6 = f'''SELECT
                count(*)
                FROM public.employers e
                left join public.employers_industries ei ON e.id = ei.employer_id 
                left join public.INDUSTRIES i on i.id = ei.industry_id
                WHERE i.name = 'Разработка программного обеспечения'
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2716940338.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [206]:
# формируем список городов миллионников
million_city = (
    "Москва", "Санкт-Петербург", "Новосибирск", 
    "Екатеринбург", "Казань", "Нижний Новгород", 
    "Красноярск", "Челябинск", "Самара", "Уфа", 
    "Ростов-на-Дону", "Омск", "Краснодар", 
    "Воронеж", "Пермь", "Волгоград"
)

# запрос для расчетаколичества вакансий компании Яндекс в городах миллионниках
query_5_7 = f'''(SELECT 
                a.name,
                count(*) cnt
                from employers e
                join vacancies v on e.id = v.employer_id
                join areas a on a.id = v.area_id
                WHERE e.name = 'Яндекс' and a.name in {million_city}
                GROUP BY a.id
                ORDER BY 2 ASC)
            
                UNION ALL 

                (SELECT 
                'TOTAL',
                COUNT (*) "Общ кол-во вакансий в городах-миллионниках"
                FROM public.vacancies v
                JOIN public.employers e ON e.id = v.employer_id
                JOIN public.areas a ON a.id = v.area_id
                WHERE e.name = 'Яндекс' and a.name in {million_city}
                order by 2)            
            ''' 
 

# выводим данные по запросу         
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2139936194.py:35: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0             TOTAL  485
1              Омск   21
2         Челябинск   22
3        Красноярск   23
4         Волгоград   24
5    Ростов-на-Дону   25
6            Казань   25
7             Пермь   25
8            Самара   26
9               Уфа   26
10        Краснодар   30
11          Воронеж   32
12      Новосибирск   35
13  Нижний Новгород   36
14     Екатеринбург   39
15  Санкт-Петербург   42
16           Москва   54

### Проведя анализ работодателей, можно сделать следующие выводы:
#### - На первом месте по количеству вакансий находится компания Яндекс (1933 вакансии), что почти в 6 раз больше чем у компании, замыкающей первую пятерку - Газпром нефть (331 вакансия) .
#### - Наибольшее количество вакансий и работодателей в Москве, наибольшее количество работодателей при отсутсвии вакансий в России. Думаю не совсем корректно указывать Россию, как отдельный регион, т.к. в  базе данные в основном разделены по городам России и СНГ, а не по странам.
#### - Наибольшее количество регионов, в котором публикуются вакансии охватывает работодатель Яндекс, на втором месте Ростелеком.
#### - У 8419 работодателей не указана сфера деятельности, думаю этим работодателям будет сложнее найти сооскателей в свои компании.
#### - Разработка программного обеспечения в качетсвке сферы деятельности указана у 3553 работодателей, т.е. у 15 % всех представленных работодателей, что показывает, что данная сфера достаточно актуальна на рынке труда. Также мы видим, что у многих работодателей указано несколько сфер деятельности.
#### - По компании Яндекс был проведен более детальный анализ. Мы видим, что данный работодатель публикует вакансии во всех исследуемых городах миллионниках, но их общее количество составляет 485 вакансий, что составляет всего 25 % от всех публикуемых вакансий компании. Из этого можно сделать вывод, что компания Яндекс привлекает большое количество соискателей не только из самых крупных регионов России, но и из более мелких регионов.



***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [207]:
# запрос для расчета количества вакансий, которые имеют отношение к данным
query_6_1 = f'''select 
                count (*)
                from public.vacancies v 
                where v.name ILIKE '%data%' OR v.name ILIKE '%данн%' OR v.name ILIKE '%Data%' OR v.name ILIKE '%Данн%'
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\906325724.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [208]:
# запрос для расчета количества вакансий для начинающего дата-сайентиста
query_6_2 = f'''select 
                count(*)
                from public.vacancies
                WHERE (name ilike '%data%scientist%' 
                    or name ilike '%data%science%'
                    or name ilike '%исследователь%данных%'
                    or name like '%ML%' and name not ilike '%HTML%'
                    or name ilike '%machine%learning%'
                    or name ilike '%машинн%обучен%%') -- выбираем вакансии для DS
                    and (name ilike '%junior%'
                    or experience = 'Нет опыта'
                    or employment = 'Стажировка') -- выбираем вакансии для начинающего DS
            '''
 

# выводим данные по запросу             
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\3098366497.py:18: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [209]:
# запрос для расчета количества вакансий с ключевым навыком SQL или postgres
query_6_3 = f'''select 
                count(*)
                from public.vacancies
                WHERE (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                and (key_skills like '%SQL%' or key_skills like 'postgres')
            '''
 

# выводим данные по запросу             
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\487552926.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [210]:
# запрос для расчета количества вакансий с ключевым навыком Python
query_6_4 = f'''select 
                count(*)
                from public.vacancies
                WHERE (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                and (key_skills ilike '%Python%')
            '''
 

# выводим данные по запросу             
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2929467289.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [211]:
# запрос для расчета среднего количества ключевых навыков, которые указывают работодатели в вакансиях для DS
query_6_5 = f'''SELECT ROUND(AVG(1 + length(key_skills) - length(Replace(key_skills, CHR(9), ''))), 2)
                FROM public.vacancies
                WHERE (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
            '''
        

# выводим данные по запросу             
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\3766434372.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [212]:
# запрос для расчета средней зарплаты для DS с опытом работы от 3 до 6 лет
query_6_6 = f'''SELECT
                ROUND(AVG(COALESCE(((salary_from + salary_to)/2), salary_from, salary_to)))
                FROM public.vacancies
                WHERE (name ilike '%data%scientist%' 
                or name ilike '%data%science%'
                or name ilike '%исследователь%данных%'
                or (name like '%ML%' and name not ilike '%HTML%')
                or name ilike '%machine%learning%'
                or name ilike '%машинн%обучен%%')
                and experience = 'От 3 до 6 лет'
                and (salary_from != 0 or salary_to != 0)
            '''
  

# выводим данные по запросу             
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\3845336025.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


round
0  243115.0

### Проведя предметный анализ вакансий, имеющих отношение к данным, можно сделать следующие выводы:
#### - В базе представлена 1771 вакансия, имеющая отношение к данным, из них 51 вакансий для начинающнго DS. Считаю, что для соискателя, не имеющего опыт работы в данной сфере, данный показатель достаточно велик, что показывает заинтересованность работодателей в специалистах, которые хотят заниматься анализом данных и развиваться в своей области.
#### - Python в качестве ключевого навыка для специалистов, работающих с данными, указывают чаще, чем SQL или postgres.
#### - В среднем работодатели в вакансиях DS указывают 6 ключевых навыков.
#### - DS с опытом работы от 3 до 6 лет может в среднем расчитывать на зарплату в 243115 рублей. Для специалиста, имеющего относительно небольшой стаж работы показатель зарплаты очень велик в сравнении с другими сферами деятельности.

***

## Дополнительные исследования

In [213]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования
# Посмотрим минимальные и макcимальные значения для каждого из двух полей зарплаты
query_7_1 = f'''SELECT 
                MIN(salary_from) min_from,
                MAX(salary_from) max_from, 
                MIN(salary_to) min_to,
                MAX(salary_to) max_to
                from public.vacancies
            '''


# выводим данные по запросу             
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\3284634563.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


min_from  max_from  min_to   max_to
0        30   1000000      60  1000000

In [214]:
# запрос, который выводит количество сфер деятельности по работодателям
query_7_2 = f'''SELECT
                e.name,
                count(ei.industry_id)
                FROM employers e 
                LEFT JOIN employers_industries ei ON e.id = ei.employer_id
                GROUP by e.name
                order by 2
             '''


# выводим данные по запросу             
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\1354344776.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


name  count
0      Савич Екатерина Юрьевна      0
1                     Автомикс      0
2             Тир Строй Монтаж      0
3                     HR Prime      0
4                    ПФ ЭРИДАН      0
...                        ...    ...
23170                    Альфа      9
23171                  Энергия     10
23172                    Прайд     10
23173          Простые решения     11
23174                   Модуль     16

[23175 rows x 2 columns]

In [215]:
# запрос, показывающий количество вакансий с удаленным режимом работы по регионам
query_7_3 = f'''SELECT
                a.name,
                count(*)
                from public.employers e
                left join vacancies v on e.id = v.employer_id
                left join areas a on a.id = v.area_id
                left join public.employers_industries ei ON e.id = ei.employer_id 
                left join public.INDUSTRIES i on i.id = ei.industry_id
                WHERE v.schedule =  'Удаленная работа'
                GROUP by a.name
                order by 2 desc
            '''
     

# выводим данные по запросу             
df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\Home\AppData\Local\Temp\ipykernel_13432\2941322770.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3, connection)


name  count
0             Москва   1751
1    Санкт-Петербург    839
2    Нижний Новгород    761
3        Новосибирск    649
4          Краснодар    648
..               ...    ...
283          Выселки      1
284      Александров      1
285         Иланский      1
286         Шарыпово      1
287     Усть-Лабинск      1

[288 rows x 2 columns]

# Общий вывод по проекту
##### - Мы провели анализ базы вакансий, представленных на HH.ru. В данной базе содержатся данные по 49197 вакансиям, 23501 работодателям, 1362 регионам России и других, 294 сферам деятельности, изучили на практическом примере применение основных принципов работы с SQL-запросами.
##### - Дополнительные исследования показывают, что в базе представлены данные с некоректно заполненными данными по зарплатной вилке. Есть вакансии с зарплатой от 30 и с зарплатой до 60. Возможно по этим вакансиям указаны иные денежные валюты. Считаю, что такие данные следует исключить из выборки при расчете средних показателей зарплаты.
##### - В представленных данных есть работодатели у которых указано 10 и более сфер деятельности (Энергия, Прайд, Простые решения, Модуль)
##### - Дополнительно проведен анализ, показывающий в каких регионах больше вакансий с удаленным режимом работы. Из него мы можем сделать вывод, что данный режим работы чаще встречается в крупных городах, а в более мелких регионах этот показатель значительно уменьшается/
###### - В дальнейшем можно было бы провести более детальные исследования, показывающие например такие взаимосвязи:
######  - соотношение регионов и среднего уровня зарплаты в них в зависимости от сферы деятельности и (или) ключевых навыков;
###### - разделить все представленные регионы по странам и провести анализ данных в зависимости от них;
###### - удалить лишние пустые и некорректные данные и провести анализ по новой базе, что позволит увидеть как "чистые" отличаются от данных, без предварительной обработки;
###### - на основе полученных данных построить графики для визуализации полученных результатов и пр.